cf/read_write/netcdf/netcdfwrite.py

```diff
@@ -455,25 +643,399 @@ def _filled_string_array(self, array, fill_value=""):
+
+        # Pad each value of the aggregation instruction arrays so that
+        # it has 'n_trailing' elements
+        a_shape = data.numblocks
+        pad = None
+        if n_trailing > 1:
+            a_shape += (n_trailing,)
+
+            # Pad the ...
+            for i, (filenames, addresses, formats) in enumerate(
+                zip(aggregation_file, aggregation_address, aggregation_format)
+            ):
+                n = n_trailing - len(filenames)
+                if n:
+                    pad = ("",) * n
+                    aggregation_file[i] = filenames + pad
+                    aggregation_format[i] = formats + pad
+                    if isinstance(addresses[0], int):
+                        pad = (-1,) * n
+
+                    aggregation_address[i] = addresses + pad
+
+        # Reshape the 1-d aggregation instruction arrays to span the
+        # data dimensions, plus the extra trailing dimension if there
+        # is one.
+        aggregation_file = np.array(aggregation_file).reshape(a_shape)
+        aggregation_address = np.array(aggregation_address).reshape(a_shape)
+        aggregation_format = np.array(aggregation_format).reshape(a_shape)
+
+        # Mask any padded elements
+        if pad:
+            aggregation_file = np.ma.where(
+                aggregation_file == "", np.ma.masked, aggregation_file
+            )
+            mask = aggregation_file.mask
+            aggregation_address = np.ma.array(aggregation_address, mask=mask)
+            aggregation_format = np.ma.array(aggregation_format, mask=mask)
+
+        # ------------------------------------------------------------
+        # Create the location array
+        # ------------------------------------------------------------
+        dtype = np.dtype(np.int32)
+        if max(data.to_dask_array().chunksize) > np.iinfo(dtype).max:
+            dtype = np.dtype(np.int64)
+
+        ndim = data.ndim
+        aggregation_location = np.ma.masked_all(
+            (ndim, max(a_shape[:ndim])), dtype=dtype
+        )
+
+        for i, chunks in enumerate(data.chunks):
+            aggregation_location[i, : len(chunks)] = chunks
+
+        # ------------------------------------------------------------
+        # Return Data objects
+        # ------------------------------------------------------------
+        data = type(data)
+        return {
+            "location": data(aggregation_location),
+            "file": data(aggregation_file),
+            "format": data(aggregation_format),
+            "address": data(aggregation_address),
+        }
+
+    def _customise_write_vars(self):
+        """Customise the write parameters.
+
+        .. versionadded:: TODOCFAVER
+
+        """
+        g = self.write_vars
+
+        if g.get("cfa"):
+            from os.path import abspath
+            from pathlib import PurePath
+
+            g["cfa_dir"] = PurePath(
+                abspath(g["filename"])
+            ).parent  # TODOCFA???
```

Comment:
Is there anything that needs doing here for this PR (RE the TODO)? Can it be updated or removed?

Reply:
No idea! Looking into it now ...

Reply:
It's all good: 5f32490b04640f3a341ab33827c287b7e64a88dd